### ISSUE TITLE
CookieAgent fails when it encounters "secure" cookies

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5175]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5175|
|Type|defect|
|Created|2011-07-03 12:26:01Z|
|Branch|https://github.com/twisted/twisted/tree/secure-cookie-5175|

Cookies can be marked as "secure" which basically means they should only be transmitted over an HTTPS connection.  `CookieAgent` passes a fake for `urllib2.Request` to the `cookielib.CookieJar` it uses, and this fails when the jar has a secure cookie and wants to know if the transport is going to be secure:

```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/exarkun/Projects/Twisted/trunk/twisted/web/client.py", line 1046, in request
    self.cookieJar.add_cookie_header(lastRequest)
  File "/usr/lib/python2.6/cookielib.py", line 1323, in add_cookie_header
    cookies = self._cookies_for_request(request)
  File "/usr/lib/python2.6/cookielib.py", line 1249, in _cookies_for_request
    cookies.extend(self._cookies_for_domain(domain, request))
  File "/usr/lib/python2.6/cookielib.py", line 1238, in _cookies_for_domain
    if not self._policy.return_ok(cookie, request):
  File "/usr/lib/python2.6/cookielib.py", line 1070, in return_ok
    if not fn(cookie, request):
  File "/usr/lib/python2.6/cookielib.py", line 1096, in return_ok_secure
    if cookie.secure and request.get_type() != "https":
AttributeError: '_FakeUrllib2Request' object has no attribute 'get_type'
```

Since `CookieAgent` is new, it would be nice to fix this before it's included in a release.

<details><summary>Searchable metadata</summary>

```
trac-id__5175 5175
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__Twisted_11_1 Twisted-11.1
branch__branches_secure_cookie_5175 branches/secure-cookie-5175
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__web web
keywords__httpclient httpclient
time__1309695961000000 1309695961000000
changetime__1310634754000000 1310634754000000
version__None None
owner__therve therve
cc__jonathanj cc__therve cc__jknight cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [32219]) Branching to 'secure-cookie-5175'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

This is ready to review in the attached branch. I spotted another problem with the port key of cookies, which I handled as well.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj set owner to @therve|
|-|-|

1.It's unfortunate that the methods of `_FakeUrllib2Request` do not have docstrings but perhaps you could add a "@see" link to <http://docs.python.org/library/urllib2.html#request-objects> to the class docstring?

 2. Looking at the source for `cookielib` there is only one method left on `urllib2.Request` that `_FakeUrllib2Request` does not implement and that is [get_origin_req_host](http://docs.python.org/library/urllib2.html#urllib2.Request.get_origin_req_host). It would be nice to finish off the implementation of `_FakeUrllib2Request`. (Though this might be more suited to the branch adding redirect support.)

 3. You're missing a newsfile.

Looks good to merge after these have been fixed and the buildbot is green.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

There's no test for the negative case. Since this is a security-related feature, it's probably doubly important to exercise the code path that causes secure cookies for a particular request to not be sent over an insecure transport (though even for non-security related cases, it's important to test positive and negative paths).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

 * I added a link to the python.org doc
 * Yeah I'll postpone that
 * Added a (empty) news file
 * I added tests for the negative case. It feels like testing cookielib, but anyway.

Back to review!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @therve|
|-|-|

Thanks for working on this, therve. The test additions look great; I have a couple of small comments if you're willing to indulge them:

 * http://tools.ietf.org/html/rfc3986 uses "scheme" and "schemes" to describe things like http and https, instead of "schema" as in the _FakeUrllib2Request docstring.
 * The use of "secure=1" in the test cookies seems to work anyway, but http://www.ietf.org/rfc/rfc2965.txt section 3.2.2 says that the "secure" attribute has no value.
 * In test_portCookie, transfered ==> transferred.

Other than that, looks great, please merge!

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/secure-cookie-5175)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [32335]) Merge secure-cookie-5175

Author: therve
Reviewers: jonathanj, exarkun, jesstess
Fixes: [#5175](https://github.com/twisted/twisted/issues/5175)

Fix the fake urllib2.Request used by CookieAgent so that secure cookies can be
used properly.
